src/engine/x86-64/X86_64Interpreter.v3

```diff
@@ -2322,6 +2322,37 @@ class X86_64InterpreterGen(ic: X86_64InterpreterCode, w: DataWriter) {
 			asm.sub_r_i(origIp, 1);
 			genDispatch0(origIp.indirect(), dispatchTables[0].1, false);
 		}
+		// specialized handler for whamm probes
+		if (FastIntTuning.enableWhammProbeTrampoline) {
+			writeDispatchEntry(dispatchTables[0].1, InternalOpcode.WHAMM_PROBE.code, w.atEnd().pos); {
+				// call runtime to get the WhammProbe object
+				computeCurIpFromIp(-1);
+				computePcFromCurIp();
+				saveCallerIVars();
+				asm.movq_r_m(r_tmp0, m_wasm_func);
+				callRuntime(refRuntimeCall(RT.runtime_GET_LOCAL_PROBE), [r_tmp0, r_curpc], false);
```

Comment:
I tried to directly emit the asm to get the probe object, but kept having memory access exceptions, couldn't figure out why the following lines won't work:

`
asm.movq_r_m(r_tmp0, m_wasm_func);
asm.movq_r_m(r_tmp1, r_tmp0.plus(offsets.WasmFunction_decl));
asm.movq_r_m(r_tmp1, r_tmp1.plus(offsets.FuncDecl_func_index));

asm.movq_r_m(r_tmp0, r_tmp0.plus(offsets.WasmFunction_instance));
asm.movq_r_m(r_tmp0, r_tmp0.plus(offsets.Instance_module));
asm.movq_r_m(r_tmp0, r_tmp0.plus(offsets.Module_probes));

asm.movq_r_m(r_tmp0, r_tmp0.plusR(r_tmp1, offsets.REF_SIZE, offsets.Array_contents));
asm.movq_r_m(r_runtime_ret0, r_tmp0.plusR(r_curpc, offsets.REF_SIZE, offsets.Array_contents));
`

Reply:
Looks like you are scaling the current PC by the `REF_SIZE` in the last line.

Reply:
but `Module` has `probes: Array<Array<Probe>>`, so I'm doing the nested array access in the last 2 lines, and the array elements are all Objects